app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/agents/AgentAccessController.scala

```diff
@@ -0,0 +1,135 @@
+import uk.gov.hmrc.cgtpropertydisposalsfrontend.controllers.SessionUpdates
+import uk.gov.hmrc.cgtpropertydisposalsfrontend.controllers.actions.{AuthenticatedAction, RequestWithSessionData, SessionDataAction, WithAuthAndSessionDataAction}
+import uk.gov.hmrc.cgtpropertydisposalsfrontend.controllers.agents.AgentAccessController._
+import uk.gov.hmrc.cgtpropertydisposalsfrontend.models.JourneyStatus.AgentStatus
+import uk.gov.hmrc.cgtpropertydisposalsfrontend.models.JourneyStatus.AgentStatus.AgentSupplyingClientDetails
+import uk.gov.hmrc.cgtpropertydisposalsfrontend.models.UserType
+import uk.gov.hmrc.cgtpropertydisposalsfrontend.models.ids.{CgtReference, GGCredId}
+import uk.gov.hmrc.cgtpropertydisposalsfrontend.repos.SessionStore
+import uk.gov.hmrc.cgtpropertydisposalsfrontend.util.Logging.LoggerOps
+import uk.gov.hmrc.cgtpropertydisposalsfrontend.util.{Logging, toFuture}
+import uk.gov.hmrc.cgtpropertydisposalsfrontend.{controllers, views}
+import uk.gov.hmrc.http.HeaderCarrier
+import uk.gov.hmrc.play.bootstrap.controller.FrontendController
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.control.NonFatal
+
+class AgentAccessController @Inject()(
+  val authenticatedAction: AuthenticatedAction,
+  val sessionDataAction: SessionDataAction,
+  authConnector: AuthConnector,
+  sessionStore: SessionStore,
+  errorHandler: ErrorHandler,
+  cc: MessagesControllerComponents,
+  enterClientsCgtRefPage: views.html.agents.enter_client_cgt_ref
+)(implicit viewConfig: ViewConfig, ec: ExecutionContext)
+    extends FrontendController(cc)
+    with WithAuthAndSessionDataAction
+    with SessionUpdates
+    with Logging { self =>
+
+  private val authorisedFunctions: AuthorisedFunctions = new AuthorisedFunctions {
+    override def authConnector: AuthConnector = self.authConnector
+  }
+
+  def enterClientsCgtRef(): Action[AnyContent] = authenticatedActionWithSessionData.async { implicit request =>
+    withAgentSupplyingClientDetails { _ =>
+      Ok(enterClientsCgtRefPage(cgtReferenceForm))
+    }
+  }
+
+  def enterClientsCgtRefSubmit(): Action[AnyContent] = authenticatedActionWithSessionData.async { implicit request =>
+    withAgentSupplyingClientDetails {
+      case AgentSupplyingClientDetails(ggCredId, _) =>
+        cgtReferenceForm
+          .bindFromRequest()
+          .fold(
+            formWithErrors => BadRequest(enterClientsCgtRefPage(formWithErrors)),
+            cgtReference => handleSubmittedCgtReferenceNumber(cgtReference, ggCredId)
+          )
+    }
+  }
+
+  private def handleSubmittedCgtReferenceNumber(
+    cgtReference: CgtReference,
+    ggCredId: GGCredId
+  )(implicit hc: HeaderCarrier, request: RequestWithSessionData[_]): Future[Result] =
+    authorisedFunctions
+      .authorised(
+        Enrolment(CgtEnrolment.enrolmentKey)
+          .withIdentifier(CgtEnrolment.enrolmentIdentifier, cgtReference.value)
+          .withDelegatedAuthRule(CgtEnrolment.delegateAuthRule)
```

Comment:
where is this rule defined? is it in the auth config?

Reply:
I think it stems from the auth config here:

https://github.com/hmrc/auth/blob/19be6274975a5031e3f07005e503ec29771e6057/conf/application.conf#L260

and that references the agent-access-control service which handles agents trying to access cgt here:

https://github.com/hmrc/agent-access-control/blob/4f54211ed074cbc6fcd6fe82563cc05033b7487e/conf/app.routes#L8